service/src/main/java/bio/terra/catalog/service/DatasetService.java

```diff
@@ -55,49 +56,76 @@ public IllegalMetadataException(Throwable cause) {
+    }
+
+    public Dataset getDataset() {
+      return dataset;
+    }
+
+    public Object convertToObject() {
+      ObjectNode node = toJsonNode(dataset.metadata());
+      node.set("accessLevel", TextNode.valueOf(String.valueOf(accessLevel)));
+      node.set("id", TextNode.valueOf(dataset.id().toValue()));
+      return node;
+    }
+
+    private ObjectNode toJsonNode(String json) {
+      try {
+        return objectMapper.readValue(json, ObjectNode.class);
+      } catch (JsonProcessingException e) {
+        // This shouldn't occur, as the data stored in postgres must be valid JSON, because it's
+        // stored as JSONB.
+        throw new IllegalMetadataException(e);
+      }
     }
   }
 
-  private List<ObjectNode> convertSourceObjectsToDatasetResponses(
+  private List<DatasetWithAccessLevel> convertSourceObjectsToDatasetsWithAccessLevel(
       Map<String, DatasetAccessLevel> roleMap, StorageSystem storageSystem) {
     List<Dataset> datasets = datasetDao.find(storageSystem, roleMap.keySet());
     return datasets.stream()
-        .map(dataset -> sourceAndDatasetToObjectNode(roleMap, dataset))
+        .map(dataset -> new DatasetWithAccessLevel(dataset, roleMap.get(dataset.storageSourceId())))
         .toList();
   }
 
-  private ObjectNode sourceAndDatasetToObjectNode(
-      Map<String, DatasetAccessLevel> roleMap, Dataset dataset) {
-    ObjectNode node = toJsonNode(dataset.metadata());
-    node.set(
-        "accessLevel", TextNode.valueOf(String.valueOf(roleMap.get(dataset.storageSourceId()))));
-    node.set("id", TextNode.valueOf(dataset.id().toValue()));
-    return node;
-  }
-
-  private List<ObjectNode> collectDatarepoDatasets(AuthenticatedUserRequest user) {
+  private List<DatasetWithAccessLevel> collectDatarepoDatasets(AuthenticatedUserRequest user) {
     // For this storage system, get the collection of visible datasets and the user's roles for
     // each dataset.
     var roleMap = datarepoService.getSnapshotIdsAndRoles(user);
-    return convertSourceObjectsToDatasetResponses(roleMap, StorageSystem.TERRA_DATA_REPO);
+    return convertSourceObjectsToDatasetsWithAccessLevel(roleMap, StorageSystem.TERRA_DATA_REPO);
   }
 
-  private List<ObjectNode> collectWorkspaceDatasets(AuthenticatedUserRequest user) {
+  private List<DatasetWithAccessLevel> collectWorkspaceDatasets(AuthenticatedUserRequest user) {
     var roleMap = rawlsService.getWorkspaceIdsAndRoles(user);
-    return convertSourceObjectsToDatasetResponses(roleMap, StorageSystem.TERRA_WORKSPACE);
+    return convertSourceObjectsToDatasetsWithAccessLevel(roleMap, StorageSystem.TERRA_WORKSPACE);
   }
 
   public DatasetsListResponse listDatasets(AuthenticatedUserRequest user) {
+    List<DatasetWithAccessLevel> datasets = new ArrayList<>();
+    datasets.addAll(collectWorkspaceDatasets(user));
+    datasets.addAll(collectDatarepoDatasets(user));
+    if (samService.hasGlobalAction(user, SamAction.READ_ANY_METADATA)) {
+      datasets.addAll(
```

Comment:
In the admin user case, we make three database calls but only one is needed. Even for the non-workspace case, we could be doing a single database call for all storage systems. When we add external dataset support we will need to revisit this anyway so there's no rush on changing it now.